artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -1183,8 +1219,126 @@ public Object getCorrelationID() {
+         if (jumpTable[i] != 0) {
+            return jumpTable;
+         }
+      }
+      // optimization over the original algorithm: it would save from accessing any jump table
+      return null;
+   }
+
+   /**
+    * https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm search algorithm
+    */
+   private static int indexOf(ReadableBuffer haystack, int start, int end, byte[] needle, int[] jumpTable) {
+      final long length = end - start;
+      int j = 0;
+      final int needleLength = needle.length;
+      for (int i = 0; i < length; i++) {
+         final int index = start + i;
+         final byte value = haystack.get(index);
+         while (j > 0 && needle[j] != value) {
+            j = jumpTable == null ? 0 : jumpTable[j];
+         }
+         if (needle[j] == value) {
+            j++;
+         }
+         if (j == needleLength) {
+            final int startMatch = index - needleLength + 1;
+            return startMatch;
+         }
+      }
+      return -1;
+   }
+
+   @Override
+   public boolean hasScheduledDeliveryTime() {
+      if (scheduledTime >= 0) {
+         return true;
+      }
+      return anyMessageAnnotations(SCHEDULED_DELIVERY_SYMBOLS,
+                                   SCHEDULED_DELIVERY_ENCODED,
+                                   SCHEDULED_DELIVERY_JUMP_TABLES);
+   }
+
+   private static boolean isValidSymbolEncoding(Symbol[] symbols, byte[][] encodedSymbols) {
+      for (int i = 0, size = symbols.length; i < size; i++) {
+         final Symbol symbol = symbols[i];
+         final byte[] encoded = encodedSymbols[i];
+         if (!Arrays.equals(symbol.toString().getBytes(StandardCharsets.US_ASCII), encoded)) {
+            return false;
+         }
+      }
+      return true;
+   }
+
+   private boolean anyMessageAnnotations(Symbol[] symbols, byte[][] encodedSymbols, int[][] jumpTables) {
+      assert isValidSymbolEncoding(symbols, encodedSymbols);
+      final int count = symbols.length;
+      if (messageDataScanned == SCANNED) {
+         final MessageAnnotations messageAnnotations = this.messageAnnotations;
+         if (messageAnnotations == null) {
+            return false;
+         }
+         Map<Symbol, Object> map = messageAnnotations.getValue();
+         if (map == null) {
+            return false;
+         }
+         for (int i = 0; i < count; i++) {
+            if (map.containsKey(symbols[i])) {
+               return true;
+            }
+         }
+         return false;
+      }
+      // messageDataScanned != SCANNED;
+      final ReadableBuffer data = this.data;
+      DecoderImpl decoder = TLSEncode.getDecoder();
+      final int position = data.position();
+      decoder.setBuffer(data.rewind());
+      try {
+         while (data.hasRemaining()) {
+            TypeConstructor<?> constructor = decoder.readConstructor();
+            if (MessageAnnotations.class.equals(constructor.getTypeClass())) {
```

Comment:
You should account for the message not having any MessageAnnotations otherwise you will scan in every type constructor for the follow on sections if present for no reason.  There's no requirement that a message have a Message Annotations section.

Reply:
How to quick search for this case? So you think it would beneficial to have such method directly on proton?

Reply:
I'm looking at 
![image](https://user-images.githubusercontent.com/13125299/74188245-5b3b6280-4c4e-11ea-9676-23e589920319.png)

If that's accurate I supposeI can use this structure to stop searching for `MessageAnnotations` earlier: any advices here is wellcome :)

Reply:
Once you pass the Header and Delivery Annotations (should they actually be there) than either you will get a Message Annotations section, or you will get one of the four possible sections that follows it or nothing if the message contains no sections following the delivery annotations.

Reply:
Great, tomorrow will add a commit to improve that + some tests.
Now my other concern is related to the lazely message scan: I've introduced a whole new state in the message and I hope it won't mess with some earlier assumptions now broken (routingType now force a message scan, while before it was assuming to always being correctly populated)

Reply:
As far as I recall the broker decodes the message annotations on each message specifically to read the routing type and delivery time on receive.  I'm unsure for routing type if that is needed on reload from journal but wouldn't be surprised if it doesn't get called at some point.

Reply:
Luckily enough on journal loading routing type seems not needed (we read it from other bits on journal record probably), while the only required message info seems the scheduled delivery time indeed

Reply:
Wouldn't it be nice if the journal entry stored all needed metadata in an easily parsed canonical format so that all protocols could benefit from the same optimizations.  :)